src/category_theory/subobject.lean

```diff
@@ -0,0 +1,1112 @@
+then pick an arbitrary choice of underlying object `(subobject.mk f : C)` back in `C`,
+it is isomorphic (in `C`) to the original `X`.
+-/
+noncomputable
+def underlying_iso {X Y : C} (f : X ⟶ Y) [mono f] : (subobject.mk f : C) ≅ X :=
+(mono_over.forget _ ⋙ over.forget _).map_iso (representative_iso (mono_over.mk' f))
+
+/--
+The morphism in `C` from the arbitrarily chosen underlying object to the ambient object.
+-/
+noncomputable
+def arrow {X : C} (Y : subobject X) : (Y : C) ⟶ X :=
+(representative.obj Y).val.hom
+
+@[simp]
+lemma underlying_arrow {X : C} {Y Z : subobject X} (f : Y ⟶ Z) :
+  underlying.map f ≫ arrow Z = arrow Y :=
+over.w (representative.map f)
+
+end subobject
+
+namespace limits
+
+section equalizer
+variables (f g : X ⟶ Y) [has_equalizer f g]
+
+/-- The equalizer of morphisms `f g : X ⟶ Y` as a `subobject X`. -/
+def equalizer_subobject : subobject X :=
+subobject.mk (equalizer.ι f g)
+
+/-- The underlying object of `equalizer_subobject f g` is (up to isomorphism!)
+the same as the chosen object `equalizer f g`. -/
+def equalizer_subobject_iso : (equalizer_subobject f g : C) ≅ equalizer f g :=
+subobject.underlying_iso (equalizer.ι f g)
+
+lemma equalizer_subobject_arrow :
+  (equalizer_subobject f g).arrow = (equalizer_subobject_iso f g).hom ≫ equalizer.ι f g :=
+(over.w (subobject.representative_iso (mono_over.mk' (equalizer.ι f g))).hom).symm
+
+@[simp] lemma equalizer_subobject_arrow' :
+  (equalizer_subobject_iso f g).inv ≫ (equalizer_subobject f g).arrow = equalizer.ι f g :=
+over.w (subobject.representative_iso (mono_over.mk' (equalizer.ι f g))).inv
+
+lemma equalizer_subobject_arrow_comp :
+  (equalizer_subobject f g).arrow ≫ f = (equalizer_subobject f g).arrow ≫ g :=
+by simp [equalizer_subobject_arrow, equalizer.condition]
+
+end equalizer
+
+section kernel
+variables [has_zero_morphisms C] (f : X ⟶ Y) [has_kernel f]
+
+/-- The kernel of a morphism `f : X ⟶ Y` as a `subobject X`. -/
+def kernel_subobject : subobject X :=
+subobject.mk (kernel.ι f)
+
+/-- The underlying object of `kernel_subobject f` is (up to isomorphism!)
+the same as the chosen object `kernel f`. -/
+def kernel_subobject_iso :
+  (kernel_subobject f : C) ≅ kernel f :=
+subobject.underlying_iso (kernel.ι f)
+
+lemma kernel_subobject_arrow :
+  (kernel_subobject f).arrow = (kernel_subobject_iso f).hom ≫ kernel.ι f :=
+(over.w (subobject.representative_iso (mono_over.mk' (kernel.ι f))).hom).symm
+
+@[simp] lemma kernel_subobject_arrow' :
+  (kernel_subobject_iso f).inv ≫ (kernel_subobject f).arrow = kernel.ι f :=
+over.w (subobject.representative_iso (mono_over.mk' (kernel.ι f))).inv
+
+lemma kernel_subobject_arrow_comp :
+  (kernel_subobject f).arrow ≫ f = 0 :=
+by simp [kernel_subobject_arrow, kernel.condition]
+
+end kernel
+
+section image
+variables (f : X ⟶ Y) [has_image f]
+
+/-- The image of a morphism `f g : X ⟶ Y` as a `subobject Y`. -/
+def image_subobject : subobject Y :=
```

Comment:
Might it be nice to have this defined as `image_mono_over`, then descend to the quotient, and build the functors `mono_over.image` and `subobject.image` out of this? I guess you didn't use the functor `subobject.image` for this since it assumes too many images like `lim.map` vs `lim_map`. If not, it's still probably a good idea to have lemmas connecting `image_subobject` and `subobject.image`

Reply:
Done.